modules/siddhi-core/src/main/java/org/wso2/siddhi/core/query/input/stream/state/AbsentLogicalPreStateProcessor.java

```diff
@@ -0,0 +1,365 @@
+            } else {
+                nextBreak = lastArrivalTime + waitingTime;
+            }
+            this.scheduler.notifyAt(nextBreak);
+        }
+    }
+
+    private boolean waitingTimePassed(long currentTime, StateEvent event) {
+        if (event.getStreamEvent(stateId) == null) {
+            // Not processed already
+            return currentTime >= event.getTimestamp() + waitingTime;
+        } else {
+            // Already processed by this processor and added back due to 'every'
+            return currentTime >= event.getStreamEvent(stateId).getTimestamp() + waitingTime;
+        }
+    }
+
+    private void sendEvent(StateEvent stateEvent) {
+        if (thisStatePostProcessor.nextProcessor != null) {
+            thisStatePostProcessor.nextProcessor.process(new ComplexEventChunk<>(stateEvent,
+                    stateEvent, false));
+        }
+        if (thisStatePostProcessor.nextStatePerProcessor != null) {
+            thisStatePostProcessor.nextStatePerProcessor.addState(stateEvent);
+        }
+        if (thisStatePostProcessor.nextEveryStatePerProcessor != null) {
+            thisStatePostProcessor.nextEveryStatePerProcessor.addEveryState(stateEvent);
+        } else if (isStartState) {
+            this.active = false;
+            if (logicalType == LogicalStateElement.Type.OR &&
+                    partnerStatePreProcessor instanceof AbsentLogicalPreStateProcessor) {
+                ((AbsentLogicalPreStateProcessor) partnerStatePreProcessor).active = false;
+            }
+        }
+        if (thisStatePostProcessor.callbackPreStateProcessor != null) {
+            thisStatePostProcessor.callbackPreStateProcessor.startStateReset();
+        }
+    }
+
+    @Override
+    public ComplexEventChunk<StateEvent> processAndReturn(ComplexEventChunk complexEventChunk) {
+
+        ComplexEventChunk<StateEvent> returnEventChunk = new ComplexEventChunk<>(false);
+
+        if (!this.active) {
+            return returnEventChunk;
+        }
+        complexEventChunk.reset();
+        StreamEvent streamEvent = (StreamEvent) complexEventChunk.next(); //Sure only one will be sent
+
+        this.lock.lock();
+        try {
+            for (Iterator<StateEvent> iterator = pendingStateEventList.iterator(); iterator.hasNext(); ) {
+                StateEvent stateEvent = iterator.next();
+                if (withinStates.size() > 0) {
+                    if (isExpired(stateEvent, streamEvent.getTimestamp())) {
+                        iterator.remove();
+                        continue;
+                    }
+                }
+                if (logicalType == LogicalStateElement.Type.OR &&
+                        stateEvent.getStreamEvent(partnerStatePreProcessor.getStateId()) != null) {
+                    iterator.remove();
+                    continue;
+                }
+                StreamEvent currentStreamEvent = stateEvent.getStreamEvent(stateId);
```

Comment:
@suhothayan 
This method has only the following section common with the parent method:
```
+            for (Iterator<StateEvent> iterator = pendingStateEventList.iterator(); iterator.hasNext(); ) {
+                StateEvent stateEvent = iterator.next();
+                if (withinStates.size() > 0) {
+                    if (isExpired(stateEvent, streamEvent.getTimestamp())) {
+                        iterator.remove();
+                        continue;
+                    }
+                }
+                if (logicalType == LogicalStateElement.Type.OR &&
+                        stateEvent.getStreamEvent(partnerStatePreProcessor.getStateId()) != null) {
+                    iterator.remove();
+                    continue;
+                }
```
the rest contsins code injections here and there. I feel spliting this into two methods and overriding will increase the complexity of the code (and reduce the readability) because the new method should receive the Iterator and some other local variables used here as parameters. Do you think it is worth to split the method in parent class and override here?